File: bench/src/swe-arena/analyze.ts
/**
 * Typed port of the reference `fixtures/analyze.py` statistics: discordant
 * pair extraction, exact two-sided sign test, and cost rollups.
 *
 * The sign test is NOT re-implemented: `mcnemar` from
 * `@tangle-network/agent-eval` computes the identical exact two-sided
 * binomial p on discordant pairs (`min(1, 2·P(X ≤ min(b,c)))`, X ~
 * Binomial(b+c, 0.5)) — the same formula as `analyze.py::sign_test`.
 *
 * Token accounting has TWO deliberately distinct rollups:
 *  - `supBrainTokens` — what analyze.py prints as "SUP total tokens":
 *    supervisor-journal metered spend (the supervisor "brain"), journal-true
 *    where captured, else the runtime's `sup_spentTokens`.
 *  - `supTotalTokens` — brain + worker-session tokens (worker-tokens.json).
 *    This is the true SUP-arm spend (1,722,390 for the 12-run set) and the
 *    number the 2.55x SUP/SOLO ratio refers to. analyze.py never printed it;
 *    its printed 0.83x ratio is brain-only.
 */

import { mcnemar } from '@tangle-network/agent-eval'
import type { ReconciledInstance } from './reconcile'
import type { LedgerRow, RematchRow, WorkerTokens } from './types'

export interface PairedOutcome {
  iid: string
  solo: boolean
  sup: boolean
}

/** Raw ledger outcomes — analyze.py semantics (strict `is True`), sorted by iid. */
export function ledgerOutcomes(ledger: LedgerRow[]): PairedOutcome[] {
  return [...ledger]
    .sort((a, b) => (a.iid < b.iid ? -1 : a.iid > b.iid ? 1 : 0))
    .map((r) => ({ iid: r.iid, solo: r.solo_resolved === true, sup: r.sup_resolved === true }))
}

/** Reconciled (re-judged, gold-gated) outcomes, sorted by iid. */
export function reconciledOutcomes(instances: ReconciledInstance[]): PairedOutcome[] {
  return [...instances]
    .sort((a, b) => (a.iid < b.iid ? -1 : a.iid > b.iid ? 1 : 0))
    .map((r) => ({ iid: r.iid, solo: r.solo.resolved, sup: r.sup.resolved }))
}

export interface DiscordantSplit {
  /** SUP resolved, SOLO not. */
  supOnly: string[]
  /** SOLO resolved, SUP not. */
  soloOnly: string[]
  both: string[]
  neither: string[]
}

export function splitPairs(outcomes: PairedOutcome[]): DiscordantSplit {
  const supOnly: string[] = []
  const soloOnly: string[] = []
  const both: string[] = []
  const neither: string[] = []
  for (const o of outcomes) {
    if (o.sup && !o.solo) supOnly.push(o.iid)
    else if (o.solo && !o.sup) soloOnly.push(o.iid)
    else if (o.solo && o.sup) both.push(o.iid)
    else neither.push(o.iid)
  }
  return { supOnly, soloOnly, both, neither }
}

export interface SignTestResult {
  n: number
  /** Discordant pairs where SUP (treatment) won. */
  b: number
  /** Discordant pairs where SOLO (control) won. */
  c: number
  nDiscordant: number
  /** Exact two-sided binomial sign-test p on discordant pairs. */
  pValue: number
}

/** Exact paired sign test, SOLO as control and SUP as treatment. */
export function pairedSignTest(outcomes: PairedOutcome[]): SignTestResult {
  const result = mcnemar(
    outcomes.map((o) => o.solo),
    outcomes.map((o) => o.sup),
  )
  const { n, b, c, nDiscordant, pValue } = result
  return { n, b, c, nDiscordant, pValue }
}

export interface CostRollup {
  soloTokens: number
  /** analyze.py's "SUP total tokens": journal-true brain spend, else sup_spentTokens. */
  supBrainTokens: number
  /** Σ worker-session tokens (worker-tokens.json; instances without workers contribute 0). */
  supWorkerTokens: number
  /** Brain + workers — the true SUP-arm token spend. */
  supTotalTokens: number
  supUsd: number
  /** Blended $ per token from SUP runtime accounting (analyze.py's `rate`). */
  blendedRatePerTok: number
  soloUsdDerived: number
  /** supBrainTokens / soloTokens — the ratio analyze.py prints (0.83x). */
  brainTokenRatio: number
  /** supTotalTokens / soloTokens — the true-spend ratio (2.55x). */
  totalTokenRatio: number
  soloWallS: number
  supWallS: number
  wallRatio: number
  /** iids where the runtime's sup_spentTokens disagrees with journal-true (no-winner zeroing). */
  telemetryGaps: string[]
}

/**
 * Port of analyze.py's cost section over the full ledger (all 12 paired runs
 * — cost is a property of the runs, not of the gradeable subset), extended
 * with the worker-token rollup.
 */
export function costRollup(
  ledger: LedgerRow[],
  supJournalTrue: Record<string, number | null>,
  workerTokens: Record<string, WorkerTokens>,
): CostRollup {
  const rows = [...ledger].sort((a, b) => (a.iid < b.iid ? -1 : a.iid > b.iid ? 1 : 0))
  let soloTokens = 0
  let supBrainTokens = 0
  let supUsd = 0
  let soloWallS = 0
  let supWallS = 0
  const telemetryGaps: string[] = []
  for (const r of rows) {
    soloTokens += r.solo_tokens
    const journalTrue = supJournalTrue[r.iid] ?? null
    supBrainTokens += journalTrue ?? r.sup_spentTokens ?? 0
    supUsd += r.sup_spentUsd ?? 0
    soloWallS += r.solo_wall_s
    supWallS += r.sup_wall_s
    if (journalTrue !== null && (r.sup_spentTokens ?? 0) !== journalTrue) telemetryGaps.push(r.iid)
  }
  const supWorkerTokens = Object.values(workerTokens).reduce((s, w) => s + w.worker_tok, 0)
  const supTotalTokens = supBrainTokens + supWorkerTokens
  const blendedRatePerTok = supBrainTokens > 0 ? supUsd / supBrainTokens : 0
  return {
    soloTokens,
    supBrainTokens,
    supWorkerTokens,
    supTotalTokens,
    supUsd,
    blendedRatePerTok,
    soloUsdDerived: soloTokens * blendedRatePerTok,
    brainTokenRatio: soloTokens > 0 ? supBrainTokens / soloTokens : 0,
    totalTokenRatio: soloTokens > 0 ? supTotalTokens / soloTokens : 0,
    soloWallS,
    supWallS,
    wallRatio: soloWallS > 0 ? supWallS / soloWallS : 0,
    telemetryGaps,
  }
}

/** Round label for the original head-to-head SUP run in the ledger. */
export type SupRound = 'SUP' | 'SUP2' | 'SUP3' | 'SUP4'

export interface RoundState {
  round: SupRound
  resolved: boolean
  verifyPass: boolean
  patchLines: number
  verdict: 'delivered' | 'no-winner' | 'best-effort' | null
  delivered: boolean | null
  spentTokens: number | null
}

/**
 * Supervisor evolution progression: the original SUP run (round 0, from the
 * ledger) followed by each rematch round, for every instance that appears in
 * at least one rematch file.
 */
export function roundsProgression(
  ledger: LedgerRow[],
  rounds: RematchRow[][],
): Map<string, RoundState[]> {
  const byIid = new Map<string, RoundState[]>()
  const rematchIids = new Set(rounds.flat().map((r) => r.iid))
  for (const r of ledger) {
    if (!rematchIids.has(r.iid)) continue
    byIid.set(r.iid, [
      {
        round: 'SUP',
        resolved: r.sup_resolved === true,
        verifyPass: r.sup_verify_pass === true,
        patchLines: r.sup_patch_lines,
        verdict: r.sup_verdict,
        delivered: r.sup_delivered,
        spentTokens: r.sup_spentTokens,
      },
    ])
  }
  for (const round of rounds) {
    for (const r of round) {
      const states = byIid.get(r.iid)
      if (!states) throw new Error(`rematch row for ${r.iid} has no ledger baseline`)
      states.push({
        round: r.arm,
        resolved: r.resolved === true,
        verifyPass: r.verify_pass === true,
        patchLines: r.patch_lines,
        verdict: r.sup_verdict,
        delivered: r.delivered,
        spentTokens: r.spentTokens,
      })
    }
  }
  return byIid
}

File: bench/src/swe-arena/fixtures/analyze.py
import json,sys,math
from itertools import combinations
HH="/tmp/claude-1000/-home-drew-code-supervisor-lab/f06fd156-042a-4ef9-bd88-f2ec7f52b90c/scratchpad/hh"
rows=[json.loads(l) for l in open(f"{HH}/ledger.jsonl") if l.strip()]
rows.sort(key=lambda r:r["iid"])
def b(x): return True if x is True else False

# --- TRUE SUP spend from journal metered events (result.spentTokens=0 on no-winner) ---
import glob as _glob
def _true_sup_tok(iid):
    js=_glob.glob(f"{HH}/runs/{iid}/SUP/ws/.loops/supervisor/*/journal.jsonl")
    if not js: return None
    ti=to=0
    for l in open(js[0]):
        try:o=json.loads(l)
        except:continue
        if o.get("kind")=="metered":
            tk=o.get("spend",{}).get("tokens",{}); ti+=tk.get("input",0) or 0; to+=tk.get("output",0) or 0
    return int(ti+to)
for r in rows:
    tt=_true_sup_tok(r["iid"])
    r["_sup_tok_true"]=tt
    r["_tok_gap"]= (tt is not None and (r.get("sup_spentTokens") or 0) != tt)

n=len(rows)
solo_res=sum(1 for r in rows if b(r.get("solo_resolved")))
sup_res=sum(1 for r in rows if b(r.get("sup_resolved")))
# discordant pairs
sup_only=[r["iid"] for r in rows if b(r.get("sup_resolved")) and not b(r.get("solo_resolved"))]
solo_only=[r["iid"] for r in rows if b(r.get("solo_resolved")) and not b(r.get("sup_resolved"))]
both=[r["iid"] for r in rows if b(r.get("sup_resolved")) and b(r.get("solo_resolved"))]
neither=[r["iid"] for r in rows if not b(r.get("sup_resolved")) and not b(r.get("solo_resolved"))]
# exact two-sided sign test on discordant pairs
def sign_test(a,b_):
    nd=a+b_
    if nd==0: return 1.0
    k=min(a,b_)
    p=sum(math.comb(nd,i) for i in range(0,k+1))/(2**nd)
    return min(1.0, 2*p)
p=sign_test(len(sup_only),len(solo_only))
print("="*80)
print(f"PAIRED HEAD-TO-HEAD: glm-5.2 SOLO vs glm-5.2 SUPERVISOR  (N={n} paired instances)")
print("="*80)
print(f"SOLO resolved: {solo_res}/{n} = {100*solo_res/n:.1f}%")
print(f"SUP  resolved: {sup_res}/{n} = {100*sup_res/n:.1f}%")
print(f"delta (SUP-SOLO): {sup_res-solo_res:+d} instances ({100*(sup_res-solo_res)/n:+.1f} pts)")
print(f"\nDISCORDANT PAIRS (the signal):")
print(f"  SUP-only wins (SUP✓ SOLO✗): {len(sup_only)}  {sup_only}")
print(f"  SOLO-only wins (SOLO✓ SUP✗): {len(solo_only)}  {solo_only}")
print(f"  both resolved: {len(both)}  |  neither: {len(neither)}  {neither}")
print(f"  exact two-sided sign test on discordant pairs: p={p:.4f}")
# cost
def s(r,k): 
    v=r.get(k); return v if isinstance(v,(int,float)) else 0
solo_tok=sum(s(r,"solo_tokens") for r in rows)
sup_tok=sum((r.get("_sup_tok_true") if r.get("_sup_tok_true") is not None else s(r,"sup_spentTokens")) for r in rows)
sup_usd=sum(s(r,"sup_spentUsd") for r in rows)
# derived solo usd at same blended rate as SUP (sup_usd/sup_tok)
rate = (sup_usd/sup_tok) if sup_tok else 0
solo_usd_derived = solo_tok*rate
print(f"\nCOST (measured tokens; USD via shared blended rate ${rate*1e6:.3f}/1M from SUP accounting):")
print(f"  SOLO total tokens: {solo_tok:,}  -> derived ${solo_usd_derived:.4f}")
print(f"  SUP  total tokens: {sup_tok:,}  -> runtime ${sup_usd:.4f}")
if solo_tok: print(f"  SUP/SOLO token ratio: {sup_tok/solo_tok:.2f}x")
if solo_usd_derived: print(f"  SUP/SOLO cost ratio (token-derived): {sup_usd/solo_usd_derived:.2f}x")
gaps=[r["iid"] for r in rows if r.get("_tok_gap")]
print(f"  [telemetry] instances where runtime spentTokens != journal-true (no-winner zeroing): {gaps}")
# wall-time cost (robust)
sw=sum(s(r,"solo_wall_s") for r in rows); pw=sum(s(r,"sup_wall_s") for r in rows)
print(f"  WALL: SOLO {sw}s total vs SUP {pw}s total -> SUP {pw/sw:.2f}x wall" if sw else "")
# per-instance table
print("\n"+"="*80); print("PER-INSTANCE"); print("="*80)
hdr=f"{'instance':32s} {'SOLO':5s} {'SUP':5s} {'v_s':3s} {'v_p':3s} {'wrk':3s} {'soloTok':8s} {'supTok':8s} {'supUSD':7s} {'soloW':5s} {'supW':5s}"
print(hdr)
for r in rows:
    print(f"{r['iid']:32s} {str(b(r.get('solo_resolved')))[:5]:5s} {str(b(r.get('sup_resolved')))[:5]:5s} "
          f"{str(b(r.get('solo_verify_pass')))[0]:3s} {str(b(r.get('sup_verify_pass')))[0]:3s} "
          f"{str(r.get('sup_workers','?')):3s} {str(s(r,'solo_tokens')):8s} {str(s(r,'sup_spentTokens')):8s} "
          f"{s(r,'sup_spentUsd'):.4f} {str(s(r,'solo_wall_s')):5s} {str(s(r,'sup_wall_s')):5s}")
print("\npatches saved under:", f"{HH}/patches/  (<iid>.solo.patch, <iid>.sup.patch)")
